pkg/render/render.go

```diff
@@ -101,11 +101,21 @@ func Agent(ctx context.Context, db kclient.Client, agent *v1.Agent, oauthServerU
 		return nil, nil, err
 	}
 
-	extraEnv, err = addKnowledgeTools(ctx, db, agent, opts.Thread, extraEnv)
+	extraEnv, added, err := addKnowledgeTools(ctx, db, agent, opts.Thread, extraEnv)
 	if err != nil {
 		return nil, nil, err
 	}
 
+	if !added {
+		filteredTools := make([]string, 0)
+		for _, tool := range mainTool.Tools {
+			if !strings.HasSuffix(tool, "as knowledge") {
```

Comment:
This is a bit hacky. Not sure if there is a great way to identify knowledge tool

Reply:
Is the part before the "as" knowledge-retrieval? I think that is the name of the tool in the tools repo.

Reply:
no, tool definition is something like this 
```
github.com/obot-platform/tools/knowledge as knowledge
```

So before `as` it is the location of the tool and last part is the tool name in index.yaml I guess

Reply:
OK. Looking at how this tool gets added to the agent, tell me if this makes sense:
1. Move the call to `configureKnowledgeEnvs` up above line 68.
2. In the loop on line 78, the `tool` that gets processed there is just called `knowledge`. That is easier to detect, and we can skip adding it to the tool if `added` is false there.
3. We would have to do that same thing for the thread loop on line 68.

The issue with this that I can think of is that we would have to do the same filtering twice now that we are adding knowledge to threads. But maybe checking for a tool called `knowledge` is better than checking if this name ends in `as knowledge`?

So, overall, I think your approach is fine and will only break if we change how the knowledge tool gets resolved.

Reply:
I second your approach @thedadams - one loop less and even less code in total :+1:

Reply:
Sounds good 👍  addressed